Sources/AppStoreConnectClient/RetryingMiddleware.swift

```diff
@@ -0,0 +1,125 @@
+    /// Delay policy for the middleware.
+    package var delay: DelayPolicy = .constant(seconds: 1)
+}
+
+extension RetryingMiddleware: ClientMiddleware {
+    /// Intercepts and potentially retries HTTP requests.
+    /// - Parameters:
+    ///   - request: The HTTP request to intercept.
+    ///   - body: The body of the HTTP request.
+    ///   - baseURL: The base URL of the request.
+    ///   - operationID: The operation ID associated with the request.
+    ///   - next: The closure representing the next middleware or endpoint in the chain.
+    /// - Returns: A tuple containing the HTTP response and body, if successful.
+    /// - Throws: An error if the request fails after all retry attempts.
+    package func intercept(
+        _ request: HTTPRequest,
+        body: HTTPBody?,
+        baseURL: URL,
+        operationID: String,
+        next: (HTTPRequest, HTTPBody?, URL) async throws -> (HTTPResponse, HTTPBody?)
+    ) async throws -> (HTTPResponse, HTTPBody?) {
+        // Check if the retry policy is set to number of attempts
+        guard case .upToAttempts(count: let maxAttemptCount) = policy else {
+            return try await next(request, body, baseURL)
+        }
+        // Check if there is a body to the request
+        if let body {
+            // Ensure the request body can be sent multiple times
+            guard body.iterationBehavior == .multiple else {
+                return try await next(request, body, baseURL)
+            }
+        }
+        // Perform an initial retry delay if necessary
+        try await willRetry()
+        // Loop for the maximum number of attempts
+        for attempt in 1...maxAttemptCount {
+            // Initialize variables for response and response body
+            let (response, responseBody): (HTTPResponse, HTTPBody?)
+            // Check if should retry after an error
+            if signals.contains(.errorThrown) {
+                do {
+                    (response, responseBody) = try await next(request, body, baseURL)
+                } catch {
+                    // If the final attempt fails, throw the error
+                    if attempt == maxAttemptCount {
+                        throw RetryingMiddlewareError.maxAttemptsReached
+                    } else {
+                        // Wait before retrying the operation
+                        try await willRetry()
+                        continue
+                    }
+                }
+            } else {
+                (response, responseBody) = try await next(request, body, baseURL)
+            }
+            // Check if the response status code indicates a retry and if max attempts are not reached
+            if signals.contains(response.status.code) && attempt < maxAttemptCount {
+                try await willRetry()
+                continue
+            } else {
+                return (response, responseBody)
+            }
+        }
```

Comment:
Add comments that describe that this case not reachable

Reply:
Added.